### ISSUE TITLE
Error in Modelica.Mechanics.MultiBody.Frames.Quaternions.Orientation.equalityConstraint

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#1817]
**Reported by hansolsson on 3 Nov 2015 10:43 UTC**
Modelica.Mechanics.MultiBody.Frames.Quaternions.Orientation.equalityConstraint seems to allow invalid ("mirrored") solutions.

E.g.:
```
Modelica.Mechanics.MultiBody.Frames.Quaternions.Orientation.equalityConstraint({1,0,0,0}, {-1,0,0,0});
# {0.0, 0.0, 0.0}
Modelica.Mechanics.MultiBody.Frames.Quaternions.orientationConstraint({1,0,0,0});
# {0.0}
Modelica.Mechanics.MultiBody.Frames.Quaternions.orientationConstraint({-1,0,0,0});
# {0.0}
```
They are generated from planarRotation({+/-1,0,0}, pi);

One solution would be (similarly to rotation matrices) to use atan2(..., Q1\*Q2). 

The "obvious" solution:
```
residue := [Q1[4], Q1[3], -Q1[2], -Q1[1]; -Q1[3], Q1[4], Q1[1], Q1[1], Q1[2], Q1[3], Q1[4]]*Q2-{0,0,1}
```
(assuming it has a unique solution) leads to an illconditioned Jacobian in combination with the orientationConstraint.

This is not good at all, but the quaternion class is not used in the connectors, which explains the low priority and high severity

I believe I earlier reported a similar issue in
Modelica.Mechanics.MultiBody.Frames.Orientation.equalityConstraint
which has now been corrected, but couldn't find the ticket-number.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1817

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 3 Nov 2015 13:47 UTC**
Replying to [ticket:1817 hansolsson]:
> I believe I earlier reported a similar issue in
> Modelica.Mechanics.MultiBody.Frames.Orientation.equalityConstraint
> which has now been corrected, but couldn't find the ticket-number.

#313 perhaps?

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 3 Nov 2015 14:24 UTC**
Replying to [comment:1 beutlich]:
> Replying to [ticket:1817 hansolsson]:
> > I believe I earlier reported a similar issue in
> > Modelica.Mechanics.MultiBody.Frames.Orientation.equalityConstraint
> > which has now been corrected, but couldn't find the ticket-number.
> #313 perhaps?

No, and it turns out that it wasn't reported on trac.modelica.org for the obvious reason that the error was from 2007-07-03 which predates trac.modelica.org by a few months.

It was still committed to svn.modelica.org.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 12 Jan 2016 16:13 UTC**
Fixed in ca2852eb8934d4044593afbbe117e62d6c1df944 as suggested

**Comment by modelica-trac-importer** [comment]
**Changelog modified by otter on 12 Jan 2016 16:13 UTC**
Quaternions equality constraints formulation improved to avoid wrong "mirror" solution
